File: java/array_search/Quickselect.java
public class Quickselect {
    private static int partition(int[] values, int left, int right) {
        int pivot = values[right];
        int store = left;

        for (int scan = left; scan < right; scan++) {
            if (values[scan] < pivot) {
                int temp = values[store];
                values[store] = values[scan];
                values[scan] = temp;
                store++;
            }
        }

        int temp = values[store];
        values[store] = values[right];
        values[right] = temp;
        return store;
    }

    public static int quickselect(int[] values, int k) {
        int left = 0;
        int right = values.length - 1;
        int target = k - 1;

        while (left <= right) {
            int pivotIndex = partition(values, left, right);

            if (pivotIndex == target) {
                return values[pivotIndex];
            }
            if (pivotIndex > target) {
                right = pivotIndex - 1;
            } else {
                left = pivotIndex + 1;
            }
        }

        return -1;
    }
}

File: c/array_search/quickselect.c
#include <stddef.h>

static void swap_int(int *left, int *right) {
    int temp = *left;
    *left = *right;
    *right = temp;
}

static size_t partition(int *values, size_t left, size_t right) {
    int pivot = values[right];
    size_t store = left;

    for (size_t scan = left; scan < right; scan++) {
        if (values[scan] < pivot) {
            swap_int(&values[store], &values[scan]);
            store++;
        }
    }

    swap_int(&values[store], &values[right]);
    return store;
}

int quickselect(int *values, size_t length, size_t k) {
    size_t left = 0;
    size_t right = length - 1;
    size_t target = k - 1;

    while (left <= right) {
        size_t pivot_index = partition(values, left, right);

        if (pivot_index == target) {
            return values[pivot_index];
        }
        if (pivot_index > target) {
            right = pivot_index - 1;
        } else {
            left = pivot_index + 1;
        }
    }

    return 0;
}
